File: F/1971-find-if-path-exists-in-graph/main.go
package main

func validPath(n int, edges [][]int, source int, destination int) bool {
	edgeMap := map[int][]int{}
	for _, edge := range edges {
		e1, e2 := edge[0], edge[1]
		_, ok := edgeMap[e1]
		if !ok {
			edgeMap[e1] = []int{}
		}
		edgeMap[e1] = append(edgeMap[e1], e2)
		_, ok2 := edgeMap[e2]
		if !ok2 {
			edgeMap[e2] = []int{}
		}
		edgeMap[e2] = append(edgeMap[e2], e1)
	}

	visited := map[int]int{}
	candis := []int{source}
	for len(candis) > 0 {
		t := []int{}
		for _, candi := range candis {
			if candi == destination {
				return true
			}
			visited[candi] = 1
			next := edgeMap[candi]
			for _, n := range next {
				_, ok := visited[n]
				if ok {
					continue
				}
				t = append(t, n)
			}
		}
		candis = t
	}
	return false
}

File: F/1971-find-if-path-exists-in-graph/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn valid_path(n: i32, edges: Vec<Vec<i32>>, source: i32, destination: i32) -> bool {
    let mut edge_map: HashMap<i32, Vec<i32>> = HashMap::new();
    for edge in edges {
      edge_map
        .entry(edge[0])
        .and_modify(|x| (*x).push(edge[1]))
        .or_insert(vec![edge[1]]);
      edge_map
        .entry(edge[1])
        .and_modify(|x| (*x).push(edge[0]))
        .or_insert(vec![edge[0]]);
    }

    let mut visited: HashMap<i32, i32> = HashMap::new();
    let mut candis: Vec<i32> = vec![source];
    while candis.len() > 0 {
      let mut t: Vec<i32> = Vec::new();
      for candi in candis {
        if candi == destination {
          return true;
        }
        visited.insert(candi, 1);
        let next = edge_map.get(&candi).unwrap();
        for n in next {
          if visited.contains_key(&n) {
            continue;
          }
          t.push(*n);
        }
      }
      candis = t;
    }
    false
  }
}
